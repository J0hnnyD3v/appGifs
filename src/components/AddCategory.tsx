import { useState } from 'react';

type Props = {
  // setCategories: React.Dispatch<React.SetStateAction<string[]>>;
  onNewCategory: (category: string) => void;
}

export const AddCategory = ({ onNewCategory }: Props) => {
  const [ inputValue, setInputValue ] = useState('');

  const onInputChange = (ev: React.ChangeEvent<HTMLInputElement>) => {
    ev.preventDefault();
    setInputValue(ev.target.value);
  };

  const onSubmit = (ev: React.FormEvent<HTMLFormElement>) => {
    ev.preventDefault();
    if (inputValue.trim().length <= 2) {
      return;
    }
    // setCategories(cat => [ inputValue, ...cat ])
    onNewCategory(inputValue.trim());
    setInputValue('');
  };

  return (
    <form onSubmit={(ev: React.FormEvent<HTMLFormElement>) => onSubmit(ev)} aria-label="form">
      <input
        type="text"
        name="category"
        id="category"
        placeholder='Buscar Gifs'
        value={inputValue}
        onChange={(ev: React.ChangeEvent<HTMLInputElement>) => onInputChange(ev)}
      />
    </form>
  );
};
