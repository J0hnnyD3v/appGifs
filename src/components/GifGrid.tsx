import { GifItem, Spinner } from './';
import { useFetchGifs } from '../hooks';

type Props = {
  category: string;
}
export const GifGrid = ({ category }: Props) => {
  const { images, isLoading } = useFetchGifs(category);

  return (
    <>
      {isLoading && (<Spinner />)}
      <h3>{category}</h3>
      <section className="card-grid">
        {images.map(img => (<GifItem key={img.id} img={img} />))}
      </section>
    </>
  );
};
