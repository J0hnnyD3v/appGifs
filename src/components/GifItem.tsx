import { Datum } from '../models/GifResponse';

type Props = {
  img: Datum
}

export const GifItem = ({ img }: Props) => {
  return (
    <article className="card">
      <p>{img.title}</p>
      <img alt={img.title} src={img.images.downsized_medium.url} />
    </article>
  );
};
