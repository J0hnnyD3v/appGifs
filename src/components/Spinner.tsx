export const Spinner = () => {
  return (
    <section className="spinner-container">
      <div className="spinner"></div>
    </section>
  );
};
