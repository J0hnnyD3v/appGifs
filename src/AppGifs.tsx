import { useState } from 'react';
import { AddCategory } from './components';
import { GifGrid } from './components';

export const AppGifs = () => {
  const [ categories, setCategories ] = useState<string[]>([ 'Dragon Ball Super', ]);

  const onAddCategory = (newCategory: string) => {
    if (categories.includes(newCategory)) {
      return;
    }
    setCategories(c => [ newCategory, ...c ]);
  };
  return (
    <>
      <h1>App Gifs</h1>

      <AddCategory
        // setCategories={setCategories}
        onNewCategory={(newCategory: string) => onAddCategory(newCategory)}
      />

      {categories.map(category => <GifGrid key={category} category={category} />)}
    </>
  );
};
