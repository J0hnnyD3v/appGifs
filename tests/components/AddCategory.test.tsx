import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';

import { AddCategory } from '../../src/components';

describe('Pruebas en <AddCategory />', () => {
  test('should match with snapshot', () => {
    const { container } = render(<AddCategory onNewCategory={() => { }} />);
    expect(container).toMatchSnapshot();
  });

  test('should change input text value', () => {
    const inputValue = 'Super Campeones';

    render(<AddCategory onNewCategory={() => { }} />);
    const input = screen.getByRole<HTMLInputElement>('textbox');

    fireEvent.input(input, { target: { value: inputValue } });
    expect(input.value).toBe(inputValue);
  });

  test('should call onNewCategory whether input text has a value', () => {
    const inputValue = 'Vegeta';
    const onNewCategory = jest.fn();

    render(<AddCategory onNewCategory={onNewCategory} />);
    const input = screen.getByRole<HTMLInputElement>('textbox');
    const form = screen.getByRole<HTMLInputElement>('form');

    fireEvent.input(input, { target: { value: inputValue } });
    fireEvent.submit(form);

    expect(input.value).toBe('');

    expect(onNewCategory).toHaveBeenCalled();
    expect(onNewCategory).toHaveBeenCalledTimes(1);
    expect(onNewCategory).toHaveBeenCalledWith(inputValue);
  });

  test('should not call onNewCategory whether input text has a short or empty value', () => {
    const onNewCategory = jest.fn();

    render(<AddCategory onNewCategory={onNewCategory} />);

    const form = screen.getByRole<HTMLInputElement>('form');

    fireEvent.submit(form);

    expect(onNewCategory).not.toHaveBeenCalled();
  });

});
