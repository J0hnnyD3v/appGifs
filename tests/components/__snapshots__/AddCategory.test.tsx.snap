// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <AddCategory /> should match with snapshot 1`] = `
<div>
  <form
    aria-label="form"
  >
    <input
      id="category"
      name="category"
      placeholder="Buscar Gifs"
      type="text"
      value=""
    />
  </form>
</div>
`;
